### PR TITLE
Improve editor popups with multiple screens

### LINKED ISSUE #37320: Windows always appear on first screen
**Godot version:**
5f1107aa3

**OS/device including version:**
Windows 10

**Issue description:**
Any new Window will appear on the first screen (monitor). IMO they should appear on the same screen as parent Window.

**Steps to reproduce:**
1. Open Godot editor
2. Move it to second screen
3. Add node
4. Dialog appears on another screen :/

EDIT:
I just noticed that any popup (including sub-menus) have this issue ;_;

### LINKED ISSUE #38591: Rightclick context menu generation uses screen offset instead of window offset.
**Alternative Title:** Rightclick context menu generation uses the active window's parent's window offset instead of active window offset.

**Godot version:**
Latest master 46f1d4ff8e111b66c457ebcf27c72f84f96721d3

**OS/device including version:**
Windows 10 1909 

**Issue description:**
Rightclick context menu generation uses screen offset instead of window offset.
Thus, even making context menu outside window depending on the position.

**Steps to reproduce:**
1. Move window anywhere accept top left default position.
2. Right-click any right-clickable thing.
3. It will be generated at the same XY offset that it should have from screen's top-left(actual) instead of window's top-left(expected).
godot-issue

**Update:** Another way to reproduce this is to right-click anything in the file dialogue e.g,
1. Click any node.
2. In the inspector, click the Load button for the Material property.
3. Move this EditorFileDialog to any position other than the default position.
4. Right-click anything.
5. **This time it will use the main window's offset instead of the EditorFileDialog's offset.**

**Minimal reproduction project:**
Any project.

### PR DESCRIPTION
**Pull request by giarve**
Hopefully this PR will be able to fix multi-screen support.

Fixes #37320
Fixes #38591

### CODE REVIEW

**Comment by kuruk-mm** [comment]
I saw your comment on IRC, I was unable to fix two screen bugs because hardware limitation (I'm trapped on Brazil because Coronavirus, and only have my notebook with limited hardware).

Set the position and size with XSizeHints is obsolete (https://tronche.com/gui/x/xlib/ICC/client-to-window-manager/wm-normal-hints.html), to set the position.

A better aproach is with https://tronche.com/gui/x/xlib/window/configure.html
But I was unable to do that for my condition. If you need to talk something you can write here or send me a IRC Private message.

**Comment by HaSa1002** [comment]
Oh sorry. I didn't saw you are working on it...

**Comment by KoBeWi** [comment]
I tested this PR and #37320 doesn't happen anymore.

**Comment by giarve** [comment]
X11 comments:

Cannot continue due to window position checks returning (0,0) sometimes and sometimes not. I have tried combining XGetWindowAttributes with XQueryTree and XTranslateCoordinates but it doesn't fix the problem. It may be the window manager being insane, but I still don't know.
Having initial coordinates at (0,0) is a problem if the function `Point2i DisplayServerX11::window_get_position(WindowID p_window)` is called before getting the real ones so that function could be replace with one that recomputes the coordinates using a working method. That race could happen in `DisplayServerX11::window_set_current_screen(int p_screen, WindowID p_window)`.

I have tried setting XConfigureWindow as @kuruk-mm (I hope you are well) suggested on the create window function but I didn't get different results. I will be pushing that structure so he won't have to make these changes.

To set a new screen for the window, the trick could be taking the current coordinates and subtracting the current screen offset (screen_get_position). Then getting the new screen offset and adding it to the previous subtraction. But we require the window position to be the real one.

X11 in conclusion: XGetWindowAttributes or XTranslateCoordinates sometimes returns incorrect positions **when the editor is not in the default/primary screen**. Watch the print_line() in __window_changed_ and the print_line() in __create_window_ while randomly clicking "Scene, Project, Debug, Editor and Help"  dropdown menus in the editor. This must be fixed to allow moving windows within screens reliably.

Due to being a problem that a lot of people will encounter I have decided to simply adapt this PR to be merged so that it is not tried to be solved multiple times.
@HaSa1002 if you're able to test your PR after this one gets merged as I tried your fix but it didn't work on Linux so you will be able to merge your changes on the gui tooltip_popup (I didn't try the tooltip tho).

This PR doesn't fix popups at all on non default/primary screens when using > 1 screens on X11.

File: platform/linuxbsd/display_server_x11.cpp [hunk 1]
```diff
@@ -829,11 +828,17 @@ void DisplayServerX11::window_set_current_screen(int p_screen, WindowID p_window
...
+		WindowData &wd = windows[p_window];
 		XMoveResizeWindow(x11_display, wd.x11_window, position.x, position.y, size.x, size.y);
 	} else {
-		if (p_screen != window_get_current_screen(p_window)) {
-			Point2i position = screen_get_position(p_screen);
-			XMoveWindow(x11_display, wd.x11_window, position.x, position.y);
+		int current_screen = window_get_current_screen(p_window);
+		if (p_screen != current_screen) {
+
+			Point2i to_screen_pos = screen_get_position(p_screen);
```
Comment (bruvzg) [inline platform/linuxbsd/display_server_x11.cpp:837]: Notes for the future improvemet:
* Special Fullscreen case probably should be ported to Windows and macOS as well.
* In case current window position (relative to the current screen) is bigger than target screen size, window will be moved to the wrong screen (not handled on any platform).

e.g. with the following display config, moving window to screen 1 will move it to screen 2 instead: Screenshot 2020-04-13 at 12 14 49

File: platform/linuxbsd/display_server_x11.cpp [hunk 2]
```diff
@@ -3352,8 +3356,9 @@ DisplayServerX11::WindowID DisplayServerX11::_create_window(WindowMode p_mode, u
 		XSync(x11_display, False);
 		XGetWindowAttributes(x11_display, wd.x11_window, &xwa);
 
-		wd.position.x = xwa.x;
-		wd.position.y = xwa.y;
+		wd.position.x = p_rect.position.x;
+		wd.position.y = p_rect.position.y;
```
Comment (bruvzg) [inline platform/linuxbsd/display_server_x11.cpp:3360]: It this correct with tiling window managers? Created window may have different size from the requested one.

**Review by bruvzg — COMMENTED** [review]
Some other places with `global_position`/`global_transform` used for pop-up:

https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/animation_bezier_editor.cpp#L654-L667
https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/connections_dialog.cpp#L828
https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/dependency_editor.cpp#L281
https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/editor_audio_buses.cpp#L506-L507
https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/editor_audio_buses.cpp#L561-L562
https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/editor_file_dialog.cpp#L605-L606
https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/editor_file_dialog.cpp#L627-L628
https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/editor_node.cpp#L4760-L4761
https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/filesystem_dock.cpp#L2267-L2268
https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/filesystem_dock.cpp#L2281-L2282
https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/filesystem_dock.cpp#L2307-L2308
https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/filesystem_dock.cpp#L2326-L2327
https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/project_settings_editor.cpp#L628-L634
https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/debugger/script_editor_debugger.cpp#L1396-L1397
https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/plugins/curve_editor_plugin.cpp#L121
https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/plugins/visual_shader_editor_plugin.cpp#L1657

These callbacks arguments probably aren't in the screen coordinates either:

https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/scene_tree_dock.cpp#L2387
https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/plugins/animation_blend_tree_editor_plugin.cpp#L452

`get_global_mouse_position()` → `get_screen_transform().xform(get_local_mouse_position()`:
https://github.com/godotengine/godot/blob/6366332949a4e9cbf327b9b19f91cc879737bdeb/editor/editor_audio_buses.cpp#L752-L753

And unlikely I have found all places.

File: platform/linuxbsd/display_server_x11.cpp (hunk 2 above)
Comment (giarve) [reply platform/linuxbsd/display_server_x11.cpp:3360]: Currently that's the only way I found it to work (at least with non tiled window managers). I tried this with i3 running together with KDE and seems to make popups like a new window just next to the editor. The size is not ok but the user can work with it I think.

However, I didn't remember to resize or move the popup (it is possible with programs like Spectacle which I used to take the photo, to move them around even if it's a tiled window manager) but it may use a special configuration (I'm not sure).

image

File: editor/plugins/script_text_editor.cpp [hunk 3]
```diff
@@ -1763,7 +1763,7 @@ void ScriptTextEditor::_make_context_menu(bool p_selection, bool p_color, bool p
 			context_menu->add_item(TTR("Pick Color"), EDIT_PICK_COLOR);
 	}
 
-	context_menu->set_position(get_global_transform().xform(p_pos));
+	context_menu->set_position(get_screen_transform().xform(p_pos));
```
[the remark this answers is not in the document]
Comment (giarve) [reply editor/plugins/script_text_editor.cpp:1766]: I added all of them and a few other ones. There may be many more left as a search for the global transform getter gave a lot of occurrences.

**Comment by giarve** [comment]
In X11, we're  currently saving the coordinates in the _create_window() function to the rect provided as argument to the function but after making a  XMoveResizeWindow() call. The problem of this is that we may not be able to set the first window position (the WM will probably place it somewhere else, like would happen with tiling WMs but it doesn't have to be tiled to ignore your placement). As a consequence, the saved coordinates may not be correct and up to date with the real position when making the window (this should only happen once). 

I would recommend to use ConfigureNotify event as a source of truth for the real position of the window, which means that only _window_changed() function should modify and save coordinates of the window structure. 

Having said that, XSync only makes sure that the command is sent, received and processed by the X server which means that it still has to give us the REAL coordinates back sending the client a ConfigureNotify event.

In conclusion, I have not found any other method to get real coordinates other than waiting for ConfigureNotify event and updating the saved coordinates accordingly (without using any type of transformation as coordinates should be global like happens in MacOS and should happen in Windows) when jumping on the callback to set the new rect in the popup class.

Future work could concentrate on waiting for the first ConfigureNotify event inside the _create_window() function or searching for another way to retrieve real coordinates (which I don't recommend if you're not an expert with X11).

File: platform/linuxbsd/display_server_x11.cpp (hunk 2 above)
Comment (giarve) [reply platform/linuxbsd/display_server_x11.cpp:3360]: Yes I have found that windows can be set to float in tiled WMs , so it shouldn't be a problem unless someone raises a problem about it :+1:

**Comment by Anutrix** [comment]
Requesting rebase.

**Comment by groud** [comment]
Sorry for the late review. It seems that we have now another PR to solve this problem: https://github.com/godotengine/godot/pull/52280. But it looks like this one might solve more situation. Would you mind a rebase?

This PR will likely need a review from @reduz, as I am not familiar enough with windows positioning.

**Comment by jmb462** [comment]
This one is more exhaustive.
It will also fix #47217 issue if merge.

**Comment by giarve** [comment]
As this PR has 10 conflicting files (which would require me to download, compile and test the code after changes or patch the code from the open PRs only and add the code changes that have been left out) I would suggest merging #52280 and #42117. 

#41395 mitigated tooltip positioning and with #40922 it is possible that X11 may work correctly after these two are merged.

However, if after merging  #52280 and #42117, someone finds that in X11 *windows* (tooltips will be probably be already fixed) are still placed in the first screen and not in the screen where the editor (which is the parent window) is running, create a new PR adding the changes from the `scene/main/window.cpp` file from this PR.

If in X11 anyone still has problems after merging these two PR mentioned above, compile the engine with both the changes from `scene/main/window.cpp` and `platform/linuxbsd/display_server_x11.cpp` and if it works, create a new PR with only these changes.

Be aware that when this PR was made, the code here was tested on at least 2 KDE desktops and an Ubuntu one (Gnome or Unity, I don't remember), working fine (tiling WMs worked but the windows or tooltips ocuppied a big screen area and I don't know if they were supposed to).

**Comment by mhilbrunner** [comment]
Alright, closing as per the above comment. :)
